ports/esp32/machine_i2s.c

```diff
@@ -0,0 +1,792 @@
+        "sck="MP_HAL_PIN_FMT ",\n"
+        "ws="MP_HAL_PIN_FMT ",\n"
+        "sd="MP_HAL_PIN_FMT ",\n"
+        "mode=%u,\n"
+        "bits=%u, format=%u,\n"
+        "rate=%d, bufferlen=%d)",
+        self->port,
+        mp_hal_pin_name(self->sck),
+        mp_hal_pin_name(self->ws),
+        mp_hal_pin_name(self->sd),
+        self->mode,
+        self->bits, self->format,
+        self->rate, self->bufferlen
+        );
+}
+
+STATIC mp_obj_t machine_i2s_make_new(const mp_obj_type_t *type, size_t n_pos_args, size_t n_kw_args, const mp_obj_t *args) {
+    mp_arg_check_num(n_pos_args, n_kw_args, 1, MP_OBJ_FUN_ARGS_MAX, true);
+
+    i2s_port_t port = mp_obj_get_int(args[0]);
+    if (port < 0 || port >= I2S_NUM_MAX) {
+        mp_raise_ValueError(MP_ERROR_TEXT("invalid id"));
+    }
+
+    machine_i2s_obj_t *self = m_new_obj(machine_i2s_obj_t);
+    self->base.type = &machine_i2s_type;
+    self->port = port;
+
+    mp_map_t kw_args;
+    mp_map_init_fixed_table(&kw_args, n_kw_args, args + n_pos_args);
+    machine_i2s_init_helper(self, n_pos_args - 1, args + 1, &kw_args);
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+STATIC mp_obj_t machine_i2s_obj_init(mp_uint_t n_pos_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    machine_i2s_init_helper(pos_args[0], n_pos_args - 1, pos_args + 1, kw_args);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(machine_i2s_init_obj, 1, machine_i2s_obj_init);
+
+STATIC mp_obj_t machine_i2s_deinit(mp_obj_t self_in) {
+    machine_i2s_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    i2s_driver_uninstall(self->port);
+
+    if (self->non_blocking_mode_task != NULL) {
+        vTaskDelete(self->non_blocking_mode_task);
+        self->non_blocking_mode_task = NULL;
+    }
+
+    if (self->non_blocking_mode_queue != NULL) {
+        vQueueDelete(self->non_blocking_mode_queue);
+        self->non_blocking_mode_queue = NULL;
+    }
+
+    self->i2s_event_queue = NULL;
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_i2s_deinit_obj, machine_i2s_deinit);
+
+STATIC mp_obj_t machine_i2s_callback(mp_obj_t self_in, mp_obj_t handler) {
+    machine_i2s_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    if (handler != mp_const_none && !mp_obj_is_callable(handler)) {
+        mp_raise_ValueError(MP_ERROR_TEXT("invalid callback"));
+    }
+
+    if (handler != mp_const_none) {
+        self->io_mode = NON_BLOCKING;
+
+        // create a queue linking the MicroPython task to a FreeRTOS task
+        // that manages the non blocking mode of operation
+        self->non_blocking_mode_queue = xQueueCreate(1, sizeof(non_blocking_descriptor_t));
+
+        // non-blocking mode requires a background FreeRTOS task
+        if (xTaskCreatePinnedToCore(
+            task_for_non_blocking_mode,
+            "i2s_non_blocking",
+            I2S_TASK_STACK_SIZE,
+            self,
+            I2S_TASK_PRIORITY,
+            (TaskHandle_t *)&self->non_blocking_mode_task,
+            MP_TASK_COREID) != pdPASS) {
```

Comment:
I wonder if there is a possibility to allow this task to run on the PRO esp32 core?

Unlike _thread  which pins new threads to the APP core there is no sharing of Micropython data and all inter task communication goes through the FreeRTOS queue.

I read some more above and my description is not quite accurate.  The **mp_buffer_into_t** passes in from the micropython script and the **non_blocking_descriptor_t** contains a pointer to it.  When reads or writes happen in the non blocking task they touch the buffer data.  

But if the callbacks are used then it reads like it should be safe because they are invoked after processing on that buffer is done.

I'll see what happens if the core is changed and report back.  

Maybe its too niche and I should just patch this in my firmware but if it works it might be a useful switch to use where you have too much processing to do and also sample gap-less audio on one esp32 core.

Reply:
I'm interested to see what happens as well.   Thanks for volunteering to try this out.

Reply:
It took me a few days to upgrade my firmware to use this new version. 

I'm trying to run the tensorflow micro_speech example and because of the machine learning model, working area, etc I don't have enough RAM to allocate 40k + 10k to buffers like your record to SD card example.

It looks like its fast enough when the task is running on the APP core for my use-case.  

I'm using 320 x 5 x 4 (6400) bytes for the i2s dma buffer and 320 x 2 (640) bytes read into buffer.  

The x 4 on the dma buffer is because your comments say it samples 32 bit integers and then throws out non-applicable byes to scale the i2s dma samples into the read into buffer.

But I also changed the task to pin to core 1 (PRO core) and it seems to run faster.   

I need to do more analysis to quantify the speedup.

When running I'm having trouble getting it to stop consistently using control-c.

If the code running is micropython then it can get caught by the except block and stop.


```
caught exception KeyboardInterrupt 
Done
Traceback (most recent call last):
  File "main.py", line 396, in <module>
  File "main.py", line 361, in runModel
  File "main.py", line 86, in input_callback
  File "micro_speech.py", line 80, in setInputTensorValues
KeyboardInterrupt: 
MicroPython v1.15-66-g7aa713ace-dirty on 2021-05-04; ESP32 module with ESP32
Type "help()" for more information.
>>> 

```

But about half the time I see this instead and the result is a soft reboot which starts up the program again.
```
caught exception KeyboardInterrupt 
Done
/opt/esp/idf/components/freertos/queue.c:764 (xQueueGenericSend)- assert failed!

abort() was called at PC 0x40093d4c on core 0

Backtrace:0x40092ca2:0x3ffce5e0 0x400933e9:0x3ffce600 0x40098db2:0x3ffce620 0x40093d4c:0x3ffce690 0x40132789:0x3ffce6d0 0x4020876b:0x3ffce710 0x4010868d:0x3ffce740 0x40101655:0x3ffce770 0x40107321:0x3ffce7a0 0x4010744e:0x3ffce7c0 0x401098a9:0x3ffce7e0 0x401016d4:0x3ffce880 0x40107321:0x3ffce8b0 0x40107409:0x3ffce8d0 0x4012fd04:0x3ffce900 0x4012fdf0:0x3ffce980 0x4012fe55:0x3ffce9a0 0x4013175f:0x3ffce9c0 0x40131a39:0x3ffcea50 0x40131a9d:0x3ffcea70 0x4011a0f0:0x3ffcea90


ELF file SHA256: 4d5c617ffc8e19ad

Rebooting...
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:5516
ho 0 tail 12 room 4
load:0x40078000,len:13800
ho 0 tail 12 room 4
load:0x40080400,len:3388
entry 0x40080634
Fatal exception (29): StoreProhibited
epc1=0x40079fd8, epc2=0x00000000, epc3=0x00000000, excvaddr=0xa2f40000, depc=0x00000000
ets Jun  8 2016 00:22:57

rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:5516
ho 0 tail 12 room 4
load:0x40078000,len:13800
ho 0 tail 12 room 4
load:0x40080400,len:3388
entry 0x40080634

```

Reply:
> It took me a few days to upgrade my firmware to use this new version.
Awesome !   Which version of ESP-IDF did you use?

> your record to SD card example

Which example are you using?   e.g.  blocking, non-blocking, uasyncio

> The x 4 on the dma buffer is because your comments say it samples 32 bit integers and then throws out non-applicable byes to scale the i2s dma samples into the read into buffer.

There is no need to be concerned with buffer ratios in the MicroPython code.  For example, you could define an internal buffer of 10k and a readbuffer of 1k,  or an internal buffer of 8k and a readbuffer of 20k.  The C code implementation shields that complexity from the user.  It was accidental that the C code comment uses 10k and 40k (which is the same as the MicroPython examples).  I will change the 10k and 40k in the C code comment to eliminate any confusion on this point.

Reply:
I setup on your i2s branch and built using the latest 4.3 idf.  I'm building using the espressif docker image: espressif/idf:release-v4.3.

I modified the non-blocking example:
https://github.com/miketeachman/micropython-i2s-examples/blob/master/examples/record-mic-to-sdcard-non-blocking.py

Thanks for working on this pull request for so long.  I think it will open up a lot of opportunities for driving LED's and various other projects that previously could only be done using C or C++.